plugins/in_docker_events/docker_events.c

```diff
@@ -75,6 +75,78 @@ static int de_unix_create(struct flb_in_de_config *ctx)
     return 0;
 }
 
+static int in_de_collect(struct flb_input_instance *ins,
+                         struct flb_config *config, void *in_context);
+
+static int reconnect_docker_sock(struct flb_input_instance *ins,
+                                 struct flb_config *config,
+                                 struct flb_in_de_config *ctx)
+{
+    int ret;
+
+    /* remove old socket collector */
+    if (ctx->coll_id >= 0) {
+        ret = flb_input_collector_delete(ctx->coll_id, ins);
+        if (ret < 0) {
+            flb_plg_error(ctx->ins, "failed to pause event");
+            return -1;
+        }
+        ctx->coll_id = -1;
+    }
+    if (ctx->fd > 0) {
+        close(ctx->fd);
+        ctx->fd = -1;
+    }
+
+    /* create socket again */
+    if (de_unix_create(ctx) < 0) {
+        flb_plg_error(ctx->ins, "failed to re-initialize socket");
+        return -1;
+    }
+    /* set event */
+    ctx->coll_id = flb_input_set_collector_event(ins,
+                                                 in_de_collect,
+                                                 ctx->fd, config);
+    if (ctx->coll_id < 0) {
+        flb_plg_error(ctx->ins,
+                      "could not set collector for IN_DOCKER_EVENTS plugin");
+        close(ctx->fd);
+        ctx->fd = -1;
+        return -1;
+    }
+    ret = flb_input_collector_start(ctx->coll_id, ins);
+    if (ret < 0) {
+        flb_plg_error(ctx->ins,
+                      "could not start collector for IN_DOCKER_EVENTS plugin");
+        flb_input_collector_delete(ctx->coll_id, ins);
+        close(ctx->fd);
+        ctx->coll_id = -1;
+        ctx->fd = -1;
+        return -1;
+    }
+    return 0;
+}
+
+static int reconnect_docker_sock_retry(struct flb_input_instance *ins,
+                                       struct flb_config *config,
+                                       struct flb_in_de_config *ctx)
+{
+    int i;
+    int ret;
+    for (i=0; i<ctx->reconnect_retry_limits; i++) {
+        ret = reconnect_docker_sock(ins, config, ctx);
+        if (ret >= 0) {
+            return ret;
+        }
+        flb_plg_info(ctx->ins, "Retry(%d) after %d seconds.",i+1,
+                     ctx->reconnect_retry_interval);
+        sleep(ctx->reconnect_retry_interval);
```

Comment:
My understanding is that this sleep() will only block a dedicated thread for this docker_events plugin, it will not block the main loop, is that correct?

If that's the case, does it simplify to do a infinite loop / waiting here?

Reply:
Oops, it blocks main loop.
We need to create timer event..